slapp/utils/merge_utils.py

```diff
@@ -0,0 +1,237 @@
+    roiLabel: str
+
+
+class Project(TypedDict):
+    sourceData: str
+    majorityLabel: str
+    workerAnnotations: List[WorkerAnnotation]
+
+
+def compute_majority(labels: list) -> Union[Any, None]:
+    """most prevalent of three labels or None if not 3
+
+    Parameters
+    ----------
+    labels: list
+        items could be for example "cell"/"not cell" or 1/0
+        our SagemakerGroundTruth jobs are using the strings
+
+    Returns
+    -------
+    majority:
+       element of labels that is most prevalent or None if
+       lenght of labels is not 3
+
+    """
+    ulabels = set(labels)
+    if len(ulabels) not in [1, 2]:
+        raise ValueError("`compute_majority()` expects binary classification "
+                         f"but received {len(ulabels)} labels: {ulabels}")
+    counts = {ulabel: labels.count(ulabel) for ulabel in ulabels}
+
+    majority = None
+
+    # require 3 and only 3 annotations
+    if sum(counts.values()) == 3:
+        mcount = max(counts.values())
+        r_counts = {v: k for k, v in counts.items()}
+        majority = r_counts[mcount]
+
+    return majority
+
+
+def get_project_key(record: dict) -> Union[str, None]:
+    """returns the key name for the project, performs some validation
+
+    Parameters
+    ----------
+    record: dict
+        the dictionary provided
+
+    Returns
+    -------
+    project_key: str
+       the key containing the project sub-dict or None if validation fails
+
+    """
+    # check that is identified
+    if 'roi-id' not in record:
+        raise MergingException("record does not contain key 'roi-id'")
+
+    # check that it conforms to
+    # https://docs.aws.amazon.com/sagemaker/latest/dg/sms-data-output.html
+    candidates = [k for k in record.keys() if '-metadata' in k]
+    if len(candidates) != 1:
+        raise MergingException(f"Record for roi-id {record['roi-id']} has "
+                               f"{len(candidates)} keys containing "
+                               "'-metadata' expecting 1 and only 1")
+
+    meta_key = candidates[0]
+    project_key = meta_key.replace("-metadata", "")
+
+    # check that there is a project entry related to the metadata
+    if project_key not in record:
+        project_key = None
+
+    return project_key
+
+
+def merge_projects(project1: Project, project2: Project) -> Project:
+    """merges worker annotations between 2 record project entries and updates
+    majorityLabel
+
+    Parameters
+    ----------
+    project1: Project
+        first project to merge
+    project2: Project
+        second project to merge
+
+    Returns
+    -------
+    new_project: Project
+        new with workerAnnotations concatenated and sourceData appended
+        if different. Any other entities inherited from project1
+
+    """
+    new_project = Project(copy.deepcopy(project1))
+    new_project['workerAnnotations'].extend(project2['workerAnnotations'])
+    if new_project['sourceData'] != project2['sourceData']:
```

Comment:
Might be clearer if this is comma-delimited rather than space, to concatenate these URIs. Did you find this case in the data?

Reply:
done.
No, but I could imagine a scenario where it would happen. Hopefully not.